app/assets/javascripts/discourse/app/lib/plugin-api/value-transformer.js

```diff
@@ -0,0 +1,214 @@
+ */
+export function _freezeValidTransformerNames() {
+  registryOpened = true;
+}
+
+/**
+ * Adds a new valid transformer name.
+ *
+ * INTERNAL API: use pluginApi.addTransformerName instead.
+ *
+ * DO NOT USE THIS FUNCTION TO ADD CORE TRANSFORMER NAMES. Instead register them directly in the
+ * validCoreTransformerNames set above.
+ *
+ * @param {string} name the name to register
+ */
+export function _addTransformerName(name) {
+  if (registryOpened) {
+    throw new Error(
+      "api.registerTransformer was called when the system is no longer accepting new names to be added.\n" +
+        `Move your code to a pre-initializer that runs before "freeze-valid-transformers" to avoid this error.`
+    );
+  }
+
+  if (validCoreTransformerNames.has(name)) {
+    // eslint-disable-next-line no-console
+    console.warn(
+      `api.addTransformerName: transformer "${name}" matches an existing core transformer and shouldn't be re-registered using the the API.`
+    );
+    return;
+  }
+
+  if (validPluginTransformerNames.has(name)) {
+    // eslint-disable-next-line no-console
+    console.warn(
+      `api.addTransformerName: transformer "${name}" is already registered.`
+    );
+    return;
+  }
+
+  validPluginTransformerNames.add(name);
+}
+
+/**
+ * Register a value transformer.
+ *
+ * INTERNAL API: use pluginApi.registerTransformer instead.
+ *
+ * @param {string} transformerName the name of the transformer
+ * @param {function({value, context})} valueCallback callback that will transform the value.
+ */
+export function _registerTransformer(transformerName, valueCallback) {
+  if (!registryOpened) {
+    throw new Error(
+      "api.registerTransformer was called while the system was still accepting new transformer names to be added.\n" +
+        `Move your code to an initializer or a pre-initializer that runs after "freeze-valid-transformers" to avoid this error.`
+    );
+  }
+
+  if (!transformerExists(transformerName)) {
+    // eslint-disable-next-line no-console
+    console.warn(
+      `api.registerTransformer: transformer "${transformerName}" is unknown and will be ignored. ` +
+        "Perhaps you misspelled it?"
+    );
+  }
+
+  if (typeof valueCallback !== "function") {
+    throw new Error(
+      "api.registerTransformer requires the valueCallback argument to be a function"
+    );
+  }
+
+  const existingTransformers = transformersRegistry.get(transformerName) || [];
+
+  existingTransformers.push(valueCallback);
+
+  transformersRegistry.set(transformerName, existingTransformers);
+}
+
+/**
+ * Apply a transformer to a value
+ *
+ * @param {string} transformerName the name of the transformer applied
+ * @param {*} defaultValue the default value
+ * @param {*} [context] the optional context to pass to the transformer callbacks.
+ *
+ * @returns {*} the transformed value
+ */
+export function applyTransformer(transformerName, defaultValue, context) {
```

Comment:
One concern I have is that people are gonna pass something weird into 'context'.

e.g. they might do

```
applyTransformer("blah", value, this);
```
and pass an entire component instance. That would make any integrations quite fragile.

Could something like this help?

```
if(context && !(typeof val === 'object' && val.constructor === Object)){
  throw "Context must be a simple JS object"
}
```

Reply:
We can add this, but TBH, it will just make it slightly harder.
If people want to pass the component in the context, they will.

It won't prevent them from doing something like:

```js
applyTransformer("blah", value, { component: this });
```

Reply:
On second thought, there is nothing we can do about third-party abusing it, but since most of these extension points will live in core and/or official plugins. 

Adding this error might just be enough to prevent us, from doing it, like a not so gentle nudge.

Reply:
Yeah agreed - it won't be perfect, but it should at least discourage it 🤞